src/tsolvers/lasolver/Simplex.cpp

```diff
@@ -0,0 +1,490 @@
+
+        explanation.push_back(itBound.getType() == bound_u ? model.readLBoundRef(it) : model.readUBoundRef(it));
+
+        explanationCoefficients.emplace_back(1);
+        explanation.push_back(itBound_ref);
+        explanationCoefficients.emplace_back(1);
+        return false;
+    }
+
+    // Update the Tableau data if a non-basic variable
+    if (tableau.isNonBasic(it)) {
+        if (!isBoundSatisfied(model.read(it), itBound)){
+            changeValueBy(it, itBound.getValue() - model.read(it));
+        }
+        else{
+//            std::cout << "Bound is satisfied by current assignment, no need to update model!\n\n";
+        }
+    }
+    else // basic variable got a new bound, it becomes a possible candidate
+    {
+        if(!tableau.isActive(it)){
+            throw "Not implemented yet!";
+        }
+        assert(tableau.isBasic(it));
+        newCandidate(it);
+    }
+
+//  LAVar *x = it;
+//  cerr << "; ASSERTED bound on " << *x << ": " << x->L( ) << " <= " << x->M( ) << " <= " << x->U( ) << endl;
+
+//  cerr  << "; NORMAL " << status <<endl;
+    return true;
+}
+
+void Simplex::newCandidate(LVRef candidateVar) {
+    assert(tableau.isBasic(candidateVar));
+    candidates.insert(candidateVar);
+}
+
+void Simplex::eraseCandidate(LVRef candidateVar) {
+    candidates.erase(candidateVar);
+}
+
+
+void Simplex::pivot(const LVRef bv, const LVRef nv){
+    assert(tableau.isBasic(bv));
+    assert(tableau.isNonBasic(nv));
+    assert(valueConsistent(bv));
+//    tableau.print();
+    updateValues(bv, nv);
+    tableau.pivot(bv, nv);
+    // after pivot, bv is not longer a candidate
+    eraseCandidate(bv);
+    // and nv can be a candidate
+    newCandidate(nv);
+//    tableau.print();
+    assert(checkTableauConsistency());
+    assert(checkValueConsistency());
+}
+
+void Simplex::changeValueBy(LVRef var, const Delta & diff) {
+    // update var's value
+    model.write(var, model.read(var) + diff);
+    // update all (active) rows where var is present
+    for ( LVRef row : tableau.getColumn(var)){
+        assert(tableau.isBasic(row));
+        if (tableau.isActive(row)) {
+            model.write(row, model.read(row) + (tableau.getCoeff(row, var) * diff));
+            newCandidate(row);
+        }
+    }
+}
+
+void Simplex::updateValues(const LVRef bv, const LVRef nv){
+    assert(model.read(bv) < model.Lb(bv) || model.read(bv) > model.Ub(bv));
+    auto bvNewVal = (model.read(bv) < model.Lb(bv)) ? model.Lb(bv) : model.Ub(bv);
+    const auto & coeff = tableau.getCoeff(bv, nv);
+    // nvDiff represents how much we need to change nv, so that bv gets to the right value
+    auto nvDiff = (bvNewVal - model.read(bv)) / coeff;
+    // update nv's value
+    changeValueBy(nv, nvDiff);
+}
+
+//
+// Returns the bounds conflicting with the actual model.
+//
+void Simplex::getConflictingBounds( LVRef x, std::vector<LABoundRef> &explanation, std::vector<opensmt::Real> &explanationCoefficients)
```

Comment:
Probably also here, it would be nice to move the out params to return value.
If the function returns something, let it return it through return value.

Reply:
That might work, also for the above thing.  It would be nice to couple together the sounds and the coefficients, when that makes sense, into a single vector that gets returned.

Reply:
Done.